### ISSUE TITLE
[dynamo][higher order ops][vmap] Recompilation on torch.compiling vmap

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#118493]
### 🐛 Describe the bug

Observed in https://github.com/pytorch/pytorch/pull/118448

~~~
import torch
import unittest
import torch._dynamo.config as config

config.capture_func_transforms = True


@torch.compile(backend="eager")
def fn(x):
    return torch.vmap(lambda x: x.sin())(x)

x = torch.zeros(3, 3, 4, 5)
y = torch.vmap(fn)(x)
with unittest.mock.patch("torch._dynamo.config.error_on_recompile", True):
    y = torch.vmap(fn)(x)
~~~

### Error logs

~~~
Traceback (most recent call last):
  File "/data/users/anijain/pytorch/examples/vma.py", line 15, in <module>
    y = torch.vmap(fn)(x)
  File "/data/users/anijain/pytorch/torch/_functorch/apis.py", line 188, in wrapped
    return vmap_impl(func, in_dims, out_dims, randomness, chunk_size, *args, **kwargs)
  File "/data/users/anijain/pytorch/torch/_functorch/vmap.py", line 281, in vmap_impl
    return _flat_vmap(
  File "/data/users/anijain/pytorch/torch/_functorch/vmap.py", line 47, in fn
    return f(*args, **kwargs)
  File "/data/users/anijain/pytorch/torch/_functorch/vmap.py", line 403, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
  File "/data/users/anijain/pytorch/torch/_dynamo/eval_frame.py", line 443, in _fn
    return fn(*args, **kwargs)
  File "/data/users/anijain/pytorch/torch/_dynamo/eval_frame.py", line 605, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
  File "/data/users/anijain/pytorch/torch/_dynamo/convert_frame.py", line 747, in _convert_frame
    result = inner_convert(frame, cache_entry, hooks, frame_state)
  File "/data/users/anijain/pytorch/torch/_dynamo/convert_frame.py", line 288, in _convert_frame_assert
    recompile_reasons = get_and_maybe_log_recompilation_reason(
  File "/data/users/anijain/pytorch/torch/_dynamo/guards.py", line 1369, in get_and_maybe_log_recompilation_reason
    raise exc.RecompileError(message)
torch._dynamo.exc.RecompileError: Recompiling function fn in /data/users/anijain/pytorch/examples/vma.py:8
    triggered by the following guard failure(s):
    - torch._C._functorch.maybe_current_level() is None             # with vmap_increment_nesting(batch_size, randomness) as vmap_level:  # _functorch/vmap.py:401 in _flat_vmap
    ~~~

### Versions

N/A

cc @ezyang @msaroufim @bdhirsh @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ydwu4

**Comment by anijain2305** [comment]
cc @zou3519

**Comment by zou3519** [comment]
cc @guilhermeleobas, seems like a bug? We shouldn't need to recompile in this situation.

**Comment by guilhermeleobas** [comment]
Closing as this was fixed in #119061
